File: trunk/java/SR-RBD/src/ddb/restore/BlockedRestoreCoordinator.java
package ddb.restore;

import ddb.Logger;
import ddb.TimeoutException;
import ddb.Worker;
import ddb.communication.TcpSender;
import ddb.msg.Message;
import ddb.msg.MessageType;
import ddb.restore.msg.RestoreNack;

public class BlockedRestoreCoordinator extends Worker {

private final static String LOGGING_NAME = "BlockedRestoreCoordinator";
	
	public void ForbidRestoration() throws InterruptedException
	{
		try 
		{
			Message msg = accept(MessageType.RESTORE_INCENTIVE, null);
			TcpSender.getInstance().sendToNode(new RestoreNack(), msg.getSender());
		} 
		catch (TimeoutException e) 
		{
			Logger.getInstance().log(
					"TimeoutException (should never happen)", 
					LOGGING_NAME, 
					Logger.Level.WARNING);
		}
	}
	
	@Override
	public void run() {
		try {
			while(true)
				ForbidRestoration();
		} catch (InterruptedException e) {
			Logger.getInstance().log(
					"InterruptedException - terminating", 
					LOGGING_NAME, 
					Logger.Level.WARNING);
		}
		
	}

}

File: C#/RBD/RBD/Restore/BlockedRestoreCoordinator.cs
using System;
using System.Threading;

using RBD.Communication;
using RBD.Msg;
using RBD.Restore.Msg;

namespace RBD.Restore {
    public class BlockedRestoreCoordinator : Worker {

    private static string LOGGING_NAME = "BlockedRestoreCoordinator";
    	
	    public void ForbidRestoration()
	    {
		    try 
		    {
			    Message msg = accept(Message.MessageType.RESTORE_INCENTIVE, null);
			    TcpSender.getInstance().sendToNode(new RestoreNack(), msg.Sender);
    		
			    Logger.getInstance().log(
					    "Restoration forbidden", 
					    LOGGING_NAME, 
					    Logger.Level.INFO);
		    } 
		    catch (TimeoutException) 
		    {
			    Logger.getInstance().log(
					    "TimeoutException (should never happen)", 
					    LOGGING_NAME, 
					    Logger.Level.WARNING);
		    }
	    }

        override public void run()
        {
		    try {
			    while(true)
				    ForbidRestoration();
		    } catch (ThreadInterruptedException) {
			    Logger.getInstance().log(
					    "InterruptedException - terminating", 
					    LOGGING_NAME, 
					    Logger.Level.WARNING);
		    }
    		
	    }

    }
}